### ISSUE TITLE
Shared config makes it hard to run behind a proxy

### ISSUE THREAD

**Issue by zeroX-tj** [askmike/gekko#672]
Hi

at the moment the shared config is a bit of a problem when running behind a proxy. I had to hardcode the port (3000) in the backend. This way i could configure 443 (ssl) for the frontend in the gekko config.

Splitting the config in a server and frontend part would solve it. What also would be cool is if we could bind the backend to an interface (localhost) so it must be accessed via the proxy.

Another thing running behind a proxy that bugged was my websocket connection getting killed. Probably because of inactivity. So i disabled that too in the fronted for now because it isn't used? Or at least i don't use it atm :) Loggers are running in seperate process, i only use the ui to run backtests. (Because i noticed loggers aren't stored and respawned on a restart?) 

I can do a PR, just would like some input first ;)

**Comment by askmike** [comment]
> I had to hardcode the port (3000) in the backend.

Ahhh yes we should split it up.

> What also would be cool is if we could bind the backend to an interface (localhost) so it must be accessed via the proxy.

👍  even better / safer than relying on users to configure a firewall properly.

> Another thing running behind a proxy that bugged was my websocket connection getting killed. 

Interesting, have not noticed this before (I assume this happened when you were not running either an import or a live gekko?). We can easily add some heartbeat in there (I thought websockets had this as part of the spec).

> So i disabled that too in the fronted for now because it isn't used?

Websockets are not used for backtests, but they are for live gekkos and imports. In the future I might use them for backtests too: if the backtest lasts very long (if you are testing over a year or so) the connection might time out.

> (Because i noticed loggers aren't stored and respawned on a restart?)

What exactly do you mean here? But nothing is stored after a backtest (simply pushed to the frontend which will discard after you backtest again).

--------

I'd love to see a PR. We should split up the ports / hosts the backend is running on VS what the frontend is trying to connect to. But I wouldn't split up the config itself: The idea is that you don't have to touch the config for normal usage (you run gekko on your own computer), and if you want to run it in the cloud it would be best to only have to edit "one" config (with a section for frontend and one for backend.

I'm kind of struggling with the config: if you don't want to use the UI you need to use gekko over the commandline with the old config, which is completely unrelated from the UIconfig (makes sense to me, but might be complex for new users - the commandline is also a "UI - User Interface").

**Comment by zeroX-tj** [comment]
> :+1: even better / safer than relying on users to configure a firewall properly.

Well, i would let the user make that decision. But still it could be a default.

>  But I wouldn't split up the config itself: The idea is that you don't have to touch the config for normal usage (you run gekko on your own computer), and if you want to run it in the cloud it would be best to only have to edit "one" config (with a section for frontend and one for backend.

I'd rather not expose information from the backend by which the user could bypass if they configured it not correctly. Anyone can read the config for the backend. You could make the default config run secure only on the user his computer. Anyone wanting to expose it as an external service can read the manual to change the interface? 

> > (Because i noticed loggers aren't stored and respawned on a restart?)

> What exactly do you mean here? But nothing is stored after a backtest (simply pushed to the frontend which will discard after you backtest again).

Sorry, i did tested the market watchers with sqlite. It was watching a market, if you then restart the UI this watcher is gone. That's why i run my loggers from cmd line (gekko is still new to me, so experimented a little with the UI and had to restart it a couple of times and set them up again) . And frankly i don't know if i'm a fan of running the backend as the main thread for everything. If that goes down, everything stops. 

I have now pm2 running as supervisor over my processes, if something goes down for the loggers it restarts them. Traders aren't restarted out of safety.  

> I'm kind of struggling with the config

So am I :laughing: 

+1 for the ws heartbeat, i did quickly try something like that, just send some json, but i never used that library and it was acting up(sometimes i could get it transmitted, sometimes not, i use the same config in other projects to proxy the upgrade connection of the websocket so the config should be ok) so i will need to investigate that again sometime.

**Comment by askmike** [comment]
> I'd rather not expose information from the backend by which the user could bypass if they configured it not correctly.

Security was never my main focus, but point taken! We do have to do it in a way that it shouldn't be that complex for users who don't want to setup a reverse proxy (run gekko on a raspberry pi in their home network).

> Sorry, i did tested the market watchers with sqlite. It was watching a market, if you then restart the UI this watcher is gone.

Yes.. At some point I want to persist them (using a db), but the biggest problem is: You want market watcher to run continuous: if you kill gekko and restart it 5 hours later, you'll end up with gaps in your market data. You cannot reliably run strats over data with gaps.

---

To sum up, I am definitely up for a PR!

**Comment by askmike** [comment]
Closing this since the PR is open at #677.
